### PR TITLE
feat: add loading indicator

### PR DESCRIPTION
**Pull request by himself65**
Before the first character responses by API, a gap in UX shows that the app is running.

Closes: https://github.com/run-llama/LlamaIndexTS/issues/200
Related: https://twitter.com/jerryjliu0/status/1724956092340613281
Related: https://twitter.com/jerryjliu0/status/1725567003501670526

https://github.com/run-llama/LlamaIndexTS/assets/14026360/62f946f4-96be-4397-aca1-2f000f70babc

### CODE REVIEW

File: packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx [hunk 1]
```diff
@@ -37,6 +38,13 @@ export default function ChatMessages(
         {props.messages.map((m) => (
           <ChatMessage key={m.id} {...m} />
         ))}
+        {props.isPending && (
+          <div
+            className='flex justify-center items-center pt-10'
+          >
+          <Spinner/>
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx:45]: how about just using `lucide-react'?
```
<Loader2 className="h-4 w-4 animate-spin" />
```

**Review by marcusschiesser — CHANGES_REQUESTED** [review]
@himself65 Great idea! Thanks for the PR. Added some comments

File: packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx [hunk 2]
```diff
@@ -23,6 +24,12 @@ export default function ChatMessages(
   const showReload =
     props.reload && !props.isLoading && isLastMessageFromAssistant;
   const showStop = props.stop && props.isLoading;
+  const content = lastMessage?.content ?? ''
+  const deferredContent = useDeferredValue(content)
+  const isPending = props.isLoading ? !isLastMessageFromAssistant : (
+    props.isLoading &&
+    isLastMessageFromAssistant && content === deferredContent
+  )
```
Comment (himself65) [inline packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx:32]: I have implemented it in another way.

I just got the idea from this section of the document (https://react.dev/reference/react/useDeferredValue#indicating-that-the-content-is-stale).

File: packages/create-llama/templates/types/streaming/nextjs/app/components/chat-section.tsx [hunk 3]
```diff
@@ -2,8 +2,12 @@
 
 import { useChat } from "ai/react";
 import { ChatInput, ChatMessages } from "./ui/chat";
+import { useCallback, useState } from 'react'
```
[the remark this answers is not in the document]
Comment (himself65) [reply packages/create-llama/templates/types/streaming/nextjs/app/components/chat-section.tsx]: Thanks for the feedback, updated

**Comment by himself65** [comment]
Just a question: How do you test the templates? I have to copy/paste from my existing project, and there's no type check (and seems no lint check) for templates in llamaindexts

**Comment by marcusschiesser** [comment]
> Just a question: How do you test the templates? I have to copy/paste from my existing project, and there's no type check (and seems no lint check) for templates in llamaindexts. Feels dangerous

current approach is to use https://github.com/run-llama/LlamaIndexTS/tree/main/packages/create-llama/templates/types/streaming and copy the UI folder

File: packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx (hunk 2 above)
Comment (marcusschiesser) [reply packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx]: I'll test this with the next release

File: packages/create-llama/templates/types/streaming/nextjs/package.json [hunk 4]
```diff
@@ -10,6 +10,7 @@
   "dependencies": {
     "ai": "^2",
     "llamaindex": "0.0.31",
+    "lucide-react": "^0.291.0",
```
Comment (marcusschiesser) [inline packages/create-llama/templates/types/streaming/nextjs/package.json:13]: don't commit changes to package.json, they are auto-generated by `create-llama`

File: packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx (hunk 2 above)
Comment (marcusschiesser) [reply packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx]: @himself65 You might also wanna have a look at https://github.com/marcusschiesser/ui - that's where the components are taken from

File: packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx [hunk 5]
```diff
@@ -23,6 +24,12 @@ export default function ChatMessages(
   const showReload =
     props.reload && !props.isLoading && isLastMessageFromAssistant;
   const showStop = props.stop && props.isLoading;
+  const content = lastMessage?.content ?? "";
+  const deferredContent = useDeferredValue(content);
+  const isPending = props.isLoading ? !isLastMessageFromAssistant : (
+    props.isLoading &&
+    isLastMessageFromAssistant && content === deferredContent
+  );
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx:32]: @himself65, this condition is a bit hard to read so a comment could help. E.g., Why do we have `props.isLoading` again in the second clause (when `props.isLoading` is `false`).

Comment (himself65) [reply packages/create-llama/templates/components/ui/shadcn/chat/chat-messages.tsx]: I thought twice and found that I should have made the logic better. updated

**Review by himself65 — COMMENTED** [review]
Btw, I found that pnpm has a different install logic than others.

For example: `typescript: ^5` will be `5.0.2` in pnpm but `5.3.2` in other package managers. This might cause some bugs in the runtime.

**Comment by marcusschiesser** [comment]
> Btw, I found that pnpm has a different install logic than others.
> 
> For example: `typescript: ^5` will be `5.0.2` in pnpm but `5.3.2` in other package managers. This might cause some bugs in the runtime.

Good catch! I had precisely the same problem yesterday. For that reason, pnpm did not work, but npm did... Would you like to look at this in one of your next PRs and fixate the version numbers?

**Review by marcusschiesser — APPROVED** [review]

**Comment by himself65** [comment]
> > Btw, I found that pnpm has a different install logic than others.
> > For example: `typescript: ^5` will be `5.0.2` in pnpm but `5.3.2` in other package managers. This might cause some bugs in the runtime.
> 
> Good catch! I had precisely the same problem yesterday. For that reason, pnpm did not work, but npm did... Would you like to look at this in one of your next PRs and fixate the version numbers?

Yes, I will check that

**Pull request merged**